numpy/core/src/multiarray/arrayfunction_override.c

```diff
@@ -329,6 +324,95 @@ array_implement_array_function(
 }
 
 
+/*
+ * Implements the __array_function__ protocol for a Python function, as described in
+ * in NEP-18. See numpy.core.overrides for a full docstring.
+ */
+NPY_NO_EXPORT PyObject *
+array_implement_array_function(
+    PyObject *NPY_UNUSED(dummy), PyObject *positional_args)
+{
+    PyObject *implementation, *public_api, *relevant_args, *args, *kwargs;
+
+    if (!PyArg_UnpackTuple(
+            positional_args, "implement_array_function", 5, 5,
+            &implementation, &public_api, &relevant_args, &args, &kwargs)) {
+        return NULL;
+    }
+
+    /* Remove `like=` kwarg, which is NumPy-exclusive and thus not present
+     * in downstream libraries.
+     */
+    if (kwargs != NULL && PyDict_Contains(kwargs, npy_ma_str_like)) {
+        PyDict_DelItem(kwargs, npy_ma_str_like);
+    }
+
+    PyObject *res = array_implement_array_function_internal(
+        public_api, relevant_args, args, kwargs);
+
+    if (res == Py_NotImplemented) {
+        return PyObject_Call(implementation, args, kwargs);
+    }
+    return res;
+}
+
+
+/*
+ * Implements the __array_function__ protocol for C array creation functions
+ * only. Added as an extension to NEP-18 in an effort to bring NEP-35 to
+ * life with minimal dispatch overhead.
+ */
+NPY_NO_EXPORT PyObject *
+array_implement_c_array_function_creation(
+    const char *function_name, PyObject *args, PyObject *kwargs)
+{
+    PyObject *result = NULL;
+
+    if (kwargs == NULL) {
+        return Py_NotImplemented;
+    }
+
+    /* Remove `like=` kwarg, which is NumPy-exclusive and thus not present
+     * in downstream libraries. If that key isn't present, return NULL and
+     * let originating call to continue.
+     */
+    if (!PyDict_Contains(kwargs, npy_ma_str_like)) {
+        return Py_NotImplemented;
+    }
+
+    PyObject *relevant_args = PyTuple_Pack(1,
+            PyDict_GetItem(kwargs, npy_ma_str_like));
+    if (relevant_args == NULL) {
+        return NULL;
+    }
+    PyDict_DelItem(kwargs, npy_ma_str_like);
+
+    PyObject *numpy_module = PyImport_Import(npy_ma_str_numpy);
+    if (numpy_module == NULL) {
+        return NULL;
+    }
+
+    PyObject *public_api = PyObject_GetAttrString(numpy_module, function_name);
+    Py_DECREF(numpy_module);
+    if (public_api == NULL) {
+        return NULL;
+    }
+    if (!PyCallable_Check(public_api)) {
+        PyErr_Format(PyExc_RuntimeError,
+                     "numpy.%s is not callable.",
+                     function_name);
+        goto cleanup;
+    }
+
+    result = array_implement_array_function_internal(
+            public_api, relevant_args, args, kwargs);
+
+cleanup:
```

Comment:
nit: could eliminate the goto here, it's only saving you a single decref
```suggestion
    if (!PyCallable_Check(public_api)) {
        PyErr_Format(PyExc_RuntimeError,
                     "numpy.%s is not callable.",
                     function_name);
        Py_DECREF(public_api);
        return NULL;
    }

    PyObject *result = array_implement_array_function_internal(
            public_api, relevant_args, args, kwargs);
```

Reply:
Not really true, note that `public_api` is passed to `array_implement_array_function_internal` and we would still need to decref that before returning.

Reply:
My patch above does not remove the decref you're referring to :)

Reply:
Done in https://github.com/numpy/numpy/pull/16935/commits/6090f3cb107caaa023b59fbbfa0f065a19c02398 .